### PR TITLE
feat: PT-4122 New trait optimized search

### PR DESCRIPTION
**Pull request by sanufriev**

### CODE REVIEW

File: enrichment/src/main/kotlin/com/rarible/protocol/union/enrichment/converter/TraitConverter.kt [hunk 1]
```diff
@@ -16,4 +22,35 @@ object TraitConverter {
...
+        )
+    }
+
+    fun convertWithRarity(traits: List<TraitEntity>, itemCount: Long): List<ExtendedTraitPropertyDto> {
+        return traits.map { trait ->
+            trait.values.map { value ->
+                ExtendedTraitPropertyDto(
+                    key = trait.key.value,
+                    value = value.value,
+                    rarity = value.count.toBigDecimal().multiply(100.toBigDecimal()).divide(itemCount.toBigDecimal(), 7, RoundingMode.HALF_UP)
```
Comment (dmitry-saraykin) [inline enrichment/src/main/kotlin/com/rarible/protocol/union/enrichment/converter/TraitConverter.kt:51]: 100.toBigDecimal() - can be constant

**Review by dmitry-saraykin — APPROVED** [review]

File: core/src/main/kotlin/com/rarible/protocol/union/core/model/elastic/EsTrait.kt [hunk 2]
```diff
@@ -9,9 +9,9 @@ data class EsTrait(
     @Id
     val id: String,
     val blockchain: BlockchainDto,
-    val collection: String?,
+    val collectionId: String,
```
Comment (yakischik) [inline core/src/main/kotlin/com/rarible/protocol/union/core/model/elastic/EsTrait.kt:12]: Is it correct to replace fields which are already in ES?

File: enrichment/src/main/kotlin/com/rarible/protocol/union/enrichment/repository/search/EsTraitRepository.kt [hunk 3]
```diff
@@ -28,4 +41,125 @@ class EsTraitRepository(
...
+        val queryBuilder = QueryBuilders.boolQuery()
+        val field = if (filter.listed) EsTrait::listedItemsCount.name else EsTrait::itemsCount.name
+        queryBuilder.must(QueryBuilders.rangeQuery(field).gt(0))
+        if (filter.collectionIds.isNotEmpty()) {
+            queryBuilder.must(QueryBuilders.termsQuery(EsTrait::collectionId.name, filter.collectionIds))
+        }
+        if (filter.keys.isNotEmpty()) {
+            queryBuilder.must(QueryBuilders.termsQuery("${EsTrait::key.name}.raw", filter.keys))
+        }
+        if (filter.text?.isNotBlank() == true) {
```
Comment (yakischik) [inline enrichment/src/main/kotlin/com/rarible/protocol/union/enrichment/repository/search/EsTraitRepository.kt:122]: You can use isBlankOrNull

File: enrichment/src/main/kotlin/com/rarible/protocol/union/enrichment/repository/search/EsTraitRepository.kt [hunk 4]
```diff
@@ -28,4 +41,125 @@ class EsTraitRepository(
...
+
+        val aggregation: Aggregation = elasticClient.aggregate(searchRequest).awaitFirstOrNull() ?: return emptyList()
+
+        return (aggregation as ParsedStringTerms).toTraitsList(field)
+    }
+
+    private fun ParsedStringTerms.toTraitsList(field: String): List<Trait> =
+        this.buckets
+            .map { keysBucket ->
+                val key = TraitEntry(value = keysBucket.keyAsString, count = keysBucket.docCount)
```
Comment (yakischik) [inline enrichment/src/main/kotlin/com/rarible/protocol/union/enrichment/repository/search/EsTraitRepository.kt:149]: Why not just single map()?

File: core/src/main/kotlin/com/rarible/protocol/union/core/model/elastic/EsTrait.kt (hunk 2 above)
Comment (sanufriev) [reply core/src/main/kotlin/com/rarible/protocol/union/core/model/elastic/EsTrait.kt:12]: collection is invalid name, in es mapping is collectionId

**Review by yakischik — APPROVED** [review]

**Pull request merged**